File: LeetCode/17.Letter_Combinations_of_a_Phone_Number/17.Letter_Combinations_of_a_Phone_Number.py
#!/usr/bin/env python3

class Solution:
	def letterCombinations(self, digits: str) -> List[str]:
		if not digits:
			return []
		
		dic = {"2":"abc", "3":"def", "4":"ghi","5":"jkl","6":"mno","7":"pqrs","8":"tuv", "9":"wxyz"}
		res = []
		
		def dfs(nums, letters):
			if not nums:
				res.append(letters)
			else:
				for a in dic[nums[0]]:
					dfs(nums[1:], letters+a)
					
		dfs(digits, "")
		return res  

File: LeetCode/17.Letter_Combinations_of_a_Phone_Number/17.Letter_Combinations_of_a_Phone_Number.java
class Solution {
    private final char[][] combinations = new char[][] { {}, {}, { 'a', 'b', 'c' }, { 'd', 'e', 'f' },
            { 'g', 'h', 'i' }, { 'j', 'k', 'l' }, { 'm', 'n', 'o' }, { 'p', 'q', 'r', 's' }, { 't', 'u', 'v' },
            { 'w', 'x', 'y', 'z' } };
    private List<String> res;

    public List<String> letterCombinations(String digits) {
        res = new ArrayList<>();
        if (digits.length() == 0) {
            return res;
        }
        backtracking(digits, new StringBuilder(), 0);
        return res;
    }

    private void backtracking(String str, StringBuilder builder, int index) {
        if (index == str.length()) {
            res.add(builder.toString());
            return;
        }

        char ch = str.charAt(index);
        char[] array = combinations[ch - '0'];
        for (char l : array) {
            builder.append(l);
            backtracking(str, builder, index + 1);
            builder.deleteCharAt(builder.length() - 1);
        }
    }
}